[IMP] sale: improve the method _create_invoices's extendability

davidtranhp opened the pull request:
It was fine before the commit a3afb98b9a90b28454284528871dd805da9a95b1

This change has no side-effect

Close #67828 




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/67829).

davidtranhp commented:
@arg-odoo Please help review. Tks!

pedrobaeza commented on addons/sale/models/sale.py:
```diff
@@ -643,7 +643,10 @@ def _create_invoices(self, grouped=False, final=False):
+            # there is a chance the invoice_vals['invoice_line_ids'] already contains data when
+            # another module extends the method `_prepare_invoice()`. Therefore, instead of
+            # replacing the invoice_vals['invoice_line_ids'], we append invoiceable lines into it
+            invoice_vals['invoice_line_ids'] += [
```
You can't do this, as this will fail if the element doesn't exist in the dictionary. Anyway, the way to do this is to append lines AFTER this method, not on `_prepare_invoice`

davidtranhp replied:
As the `_prepare_invoice()` always gives the `invoice_line_ids` as `[]`. It's safe to apply the operator `+=`
However, do you think I should test if it exists and is a list first?

davidtranhp replied:
FYI, In my situation, a single sales order line may generate more than one invoice lines. With the current code structure, I have no way to add more invoice lines than except overriding the `_prepare_invoice()`

pedrobaeza reviewed (changes requested):
Although is good to have it, you are intercepting calls in the wrong place for what you want to do, so for me this is not needed, as you have another way to do it.

Feyensv reviewed (commented):
Well, I don't really get when you would want multiple invoice lines for a SO line, but I recognize it was a slight behavior change in stable and therefore I don't mind bringing back the old behavior.

davidtranhp commented on addons/sale/models/sale.py:
```diff
@@ -643,7 +643,10 @@ def _create_invoices(self, grouped=False, final=False):
             if not invoiceable_lines:
```
Tks for your suggestion. Done now.

Feyensv commented:
Please also update your commit message, changes targeting stable branches are expected to be [FIX] commits, and should explain clearly the problem solved by the commit :).

```
[FIX] sale: do not overwrite invoice values specified by `_prepare_invoice`

Since a3afb98, the `invoice_line_ids` value is replaced in `_create_invoice`.
This broke some custom client overrides based on overrides of `_prepare_invoice`.

This commit brings back the old behavior, ensuring that the `invoice_line_ids` values 
specified by `_prepare_invoice` are extended but not replaced.
```

davidtranhp commented:
> Please also update your commit message, changes targeting stable branches are expected to be [FIX] commits, and should explain clearly the problem solved by the commit :).
> 
> ```
> [FIX] sale: do not overwrite invoice values specified by `_prepare_invoice`
> 
> Since a3afb98, the `invoice_line_ids` value is replaced in `_create_invoice`.
> This broke some custom client overrides based on overrides of `_prepare_invoice`.
> 
> This commit brings back the old behavior, ensuring that the `invoice_line_ids` values 
> specified by `_prepare_invoice` are extended but not replaced.
> ```

Tks. I have changed it and rebased + force push.

Feyensv reviewed (approved):
@robodoo r+

Yenthe666 commented:
Hah, we where cought by this one today too. Took me a bit to find back this fix :) 
@Feyensv let me give you a functional sample. You sell a shoe to a customer and you also have a mobile app in which you can gain virtual points. These virtual points can be traded in for euro's (so discount). However, on the order you want the actual total and on the invoice (for accounting reasons) you want a deducation line for the virtual credits of -xx euro.
